### PR TITLE
Python 3.9 module for RHEL8

### PR DESCRIPTION
**Pull request by frenzymadness**
The change in the first commit does not have any impact on the generated files but makes it easier to add a new version of Python, where we need only standard pythonX.Y{,-devel} "extra" packages.

I'm gonna test the new image internally.

### CODE REVIEW

**Review by torsava — APPROVED** [review]
LGTM! Thanks for thinking about simplifying the definitions!

**Comment by frenzymadness** [comment]
Fix for tests: #426 

Fixed ubi versions and the main image URL.

**Comment by torsava** [comment]
> Fix for tests: #426

Left a comment there.

> Fixed ubi versions and the main image URL.

Looks good!

**Comment by frenzymadness** [comment]
[test][test-openshift]

**Comment by frenzymadness** [comment]
[test][test-openshift]

**Comment by frenzymadness** [comment]
[test][test-openshift]

**Comment by frenzymadness** [comment]
[test][test-openshift]

**Comment by phracek** [comment]
The error looks like Jenkins CI problem https://github.com/sclorg/ci-scripts/pull/34.
As soon as it is merged, Let's start tests again.

**Comment by phracek** [comment]
Let's try one more time. CI is fixed.

[test][test-openshift]

**Comment by phracek** [comment]
Let's try one more time. CI is fixed.

[test][test-openshift]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
Failure for RHEL

```
STEP 7: RUN INSTALL_PKGS="python39 python39-devel python39-setuptools python39-pip nss_wrapper         httpd httpd-devel mod_ssl mod_auth_gssapi mod_ldap         mod_session atlas-devel gcc-gfortran libffi-devel libtool-ltdl enchant" &&     yum -y module enable python39:3.9 httpd:2.4 &&     yum -y --setopt=tsflags=nodocs install $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     rpm -e --nodeps redhat-logos-httpd &&     yum -y clean all --enablerepo='*'
Updating Subscription Management repositories.
Unable to read consumer identity
Subscription Manager is operating in container mode.
Red Hat Enterprise Linux 8 for x86_64 - AppStre  33 MB/s |  26 MB     00:00    
Red Hat Enterprise Linux 8 for x86_64 - BaseOS   33 MB/s |  28 MB     00:00    
Red Hat Universal Base Image 8 (RPMs) - BaseOS  991 kB/s | 777 kB     00:00    
Red Hat Universal Base Image 8 (RPMs) - AppStre 5.1 MB/s | 5.2 MB     00:01    
Red Hat Universal Base Image 8 (RPMs) - CodeRea  29 kB/s |  15 kB     00:00    
Error: Problems in request:
missing groups or modules: python39:3.9
Error: error building at STEP "RUN INSTALL_PKGS="python39 python39-devel python39-setuptools python39-pip nss_wrapper         httpd httpd-devel mod_ssl mod_auth_gssapi mod_ldap         mod_session atlas-devel gcc-gfortran libffi-devel libtool-ltdl enchant" &&     yum -y module enable python39:3.9 httpd:2.4 &&     yum -y --setopt=tsflags=nodocs install $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     rpm -e --nodeps redhat-logos-httpd &&     yum -y clean all --enablerepo='*'": error while running runtime: exit status 1
make[1]: Leaving directory '/root/sources'
make[1]: *** [common/common.mk:87: 3.9] Error 125
make: *** [common/common.mk:76: build-serial] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by frenzymadness** [comment]
The failure is expected because the module is not released yet. Now I'm thinking about postponing this PR until the next RHEL is released to prevent breaking CI with Python 3.9. @torsava what do you think?

**Comment by torsava** [comment]
> The failure is expected because the module is not released yet. Now I'm thinking about postponing this PR until the next RHEL is released to prevent breaking CI with Python 3.9. @torsava what do you think?

Agreed. If somebody really wants a python39 container right now, they can open a PR to add it from CentOS Stream 8.

**Comment by phracek** [comment]
Failure for RHEL

```
STEP 7: RUN INSTALL_PKGS="python39 python39-devel python39-setuptools python39-pip nss_wrapper         httpd httpd-devel mod_ssl mod_auth_gssapi mod_ldap         mod_session atlas-devel gcc-gfortran libffi-devel libtool-ltdl enchant" &&     yum -y module enable python39:3.9 httpd:2.4 &&     yum -y --setopt=tsflags=nodocs install $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     rpm -e --nodeps redhat-logos-httpd &&     yum -y clean all --enablerepo='*'
Updating Subscription Management repositories.
Unable to read consumer identity
Subscription Manager is operating in container mode.
Red Hat Enterprise Linux 8 for x86_64 - AppStre  33 MB/s |  26 MB     00:00    
Red Hat Enterprise Linux 8 for x86_64 - BaseOS   33 MB/s |  28 MB     00:00    
Red Hat Universal Base Image 8 (RPMs) - BaseOS  991 kB/s | 777 kB     00:00    
Red Hat Universal Base Image 8 (RPMs) - AppStre 5.1 MB/s | 5.2 MB     00:01    
Red Hat Universal Base Image 8 (RPMs) - CodeRea  29 kB/s |  15 kB     00:00    
Error: Problems in request:
missing groups or modules: python39:3.9
Error: error building at STEP "RUN INSTALL_PKGS="python39 python39-devel python39-setuptools python39-pip nss_wrapper         httpd httpd-devel mod_ssl mod_auth_gssapi mod_ldap         mod_session atlas-devel gcc-gfortran libffi-devel libtool-ltdl enchant" &&     yum -y module enable python39:3.9 httpd:2.4 &&     yum -y --setopt=tsflags=nodocs install $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     rpm -e --nodeps redhat-logos-httpd &&     yum -y clean all --enablerepo='*'": error while running runtime: exit status 1
make[1]: Leaving directory '/root/sources'
make[1]: *** [common/common.mk:87: 3.9] Error 125
make: *** [common/common.mk:76: build-serial] Error 2
Build step 'Execute shell' marked build as failure
```

> The failure is expected because the module is not released yet. Now I'm thinking about postponing this PR until the next RHEL is released to prevent breaking CI with Python 3.9. @torsava what do you think?

What about adding into generated section file .exclude-rhel8 which will suppress the tests until the module is not released?

**Comment by torsava** [comment]
> > The failure is expected because the module is not released yet. Now I'm thinking about postponing this PR until the next RHEL is released to prevent breaking CI with Python 3.9. @torsava what do you think?
> 
> What about adding into generated section file .exclude-rhel8 into 3.9 version, which will suppress the tests until the module is not released?

So can we even build the python-39 container here on github right now?

Another argument for postponing the PR is that we mention `registry.access.redhat.com/ubi8/python-39` in the documentation quite a few times, and the link won't be working until the GA.

**Comment by frenzymadness** [comment]
I'd rather wait. It's easier to forget to remove the exclude file than to merge an open PR. We already have Python 3.9 container based on Fedora.

**Comment by frenzymadness** [comment]
Rebased and once more verified that the Dockerfile is identical in downstream. RHEL 8.4 is out with Python 3.9 module and images so I'm gonna merge this if the CI will be at least partially green.

[test][test-openshift][test-openshift-4]

**Comment by frenzymadness** [comment]
The test result for RHEL 8 seems to be incomplete for some reason, let's try it once more [test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
RHEL8 failed in test: `test_from_dockerfile`

```bash
  File "/usr/lib64/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/opt/app-root/lib64/python3.8/site-packages/pip/_vendor/urllib3/response.py", line 443, in _error_catcher
    raise ReadTimeoutError(self._pool, None, "Read timed out.")
pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.
Error: error building at STEP "RUN pip install -U "pip>=19.3.1" &&     pip install -r requirements.txt &&     python manage.py collectstatic --noinput &&     python manage.py migrate": error while running runtime: exit status 2
ERROR: The image cannot be built from /root/sources/3.8/test/from-dockerfile/Dockerfile and application https://github.com/sclorg/django-ex.git@2.2.x.
Terminating the Dockerfile build.
```

**Review by phracek — APPROVED** [review]
LGTM.

**Pull request merged**
